flake.nix

```diff
@@ -0,0 +1,169 @@
+{
+  description = "High-level tracing language for Linux eBPF";
+
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/release-22.11";
+    flake-utils.url = "github:numtide/flake-utils";
+  };
+
+  outputs = { self, nixpkgs, flake-utils, ... }:
+    # We only support 64-bit linux systems
+    flake-utils.lib.eachSystem [ "x86_64-linux" "aarch64-linux" ]
+      (system:
+        let
+          # Overlay to specify build should use the specific libbpf we want
+          libbpfVersion = "1.2.0";
+          libbpfOverlay =
+            (self: super: {
+              libbpf_1 = super.libbpf_1.overrideAttrs (old: {
+                version = libbpfVersion;
+                src = super.fetchFromGitHub {
+                  owner = "libbpf";
+                  repo = "libbpf";
+                  rev = "v${libbpfVersion}";
+                  # If you don't know the hash the first time, set:
+                  # hash = "";
+                  # then nix will fail the build with such an error message:
+                  # hash mismatch in fixed-output derivation '/nix/store/m1ga09c0z1a6n7rj8ky3s31dpgalsn0n-source':
+                  # specified: sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=
+                  # got:    sha256-173gxk0ymiw94glyjzjizp8bv8g72gwkjhacigd1an09jshdrjb4
+                  sha256 = "sha256-NimK4pdYcai21hZHdP1mBX1MOlNY61iDJ+PDYwpRuVE=";
+                };
+              });
+            });
+
+          # Overlay to specify build should use the specific bcc we want
+          bccVersion = "0.27.0";
+          bccOverlay =
+            (self: super: {
+              bcc = super.bcc.overridePythonAttrs (old: {
+                version = bccVersion;
+                src = super.fetchFromGitHub {
+                  owner = "iovisor";
+                  repo = "bcc";
+                  rev = "v${bccVersion}";
+                  sha256 = "sha256-+RK5RZcoNHlgMOFPgygRf2h+OZGxR9gJ+fTbYjDB6Ww=";
+                };
+                # Seems like these extra tools are needed to build bcc
+                nativeBuildInputs = old.nativeBuildInputs ++ [ pkgs.python310Packages.setuptools pkgs.zip ];
+              });
+            });
+
+          # We need to use two overlays so that bcc inherits the our pinned libbpf
+          pkgs = import nixpkgs { inherit system; overlays = [ libbpfOverlay bccOverlay ]; };
```

Comment:
Do we need these overlays? It appears to at least build fine with the current versions in the `nixpkgs` referred to by the `flake.lock` file. I agree that they may be needed at some point in the future so perhaps it is better to have the infrastructure, and I understand the motivation to keep them synced with the submodules. Worth a discussion at least.

Reply:
Yes, I think we need these. Based on the discussion in #2334 and #2322 there is consensus that bpftrace should have a strong dependency on libbpf, especially to ease development and migration away from bcc. It is too difficult to support every libbpf under the sun and perform possibly the largest refactor this project has seen. During release time we will simply bump the required version to the one we developed against (which solve the packaging problem too).

Reply:
I think you would still have precise control as there is control over when the lock and hence `nixpkgs` dependency is updated, but this definitely makes it more explicit. Fine by me.